### PR TITLE
Added auth module for JWT based authentication. Fixed dependencies fo…

### PR DESCRIPTION
**Pull request by DanielTheNipapHacker**
…r Docker development setup.

### CODE REVIEW

File: nipap/nipap/authlib.py [hunk 1]
```diff
@@ -271,6 +287,107 @@ def authorize(self):
...
+            # Auth failed
+            self._logger.debug('could not authenticate')
+            self._authenticated = False
+            return self._authenticated
+
+        # auth succeeded
+        if self._authenticated:
+            self.username = payload.get('sub')
+            self.authenticated_as = payload.get('sub')
+            self.full_name = payload.get('sub')
```
Comment (garberg) [inline nipap/nipap/authlib.py:385]: There is a standardized claim called "name" which as I understand it is meant to contain the full name of the user. Could we use that one instead?

File: nipap/nipap/authlib.py [hunk 2]
```diff
@@ -271,6 +287,107 @@ def authorize(self):
...
+            # Fetch JWKs (done when initializing JwtAuth-class),
+            # keep the keys in the class instance
+            jwk_request_response = requests.get(self._token)
+            jwks = jwk_request_response.json()
+            jwk_keys = {}
+            for jwk in jwks['keys']:
+                kid = jwk['kid']
+                jwk_keys[kid] = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(jwk))
+
+            # # Upon auth with a JWT-token
```
Comment (garberg) [inline nipap/nipap/authlib.py:355]: Double-commented? :)

File: nipap/nipap/authlib.py [hunk 3]
```diff
@@ -271,6 +287,103 @@ def authorize(self):
...
+
+            # # Setting read and write rights
+            if payload.get('groups'):
+                if self._jwt_ro_group in payload.get('groups'):
+                    self.readonly = True
+                    self._authenticated = True
+                elif self._jwt_rw_group in payload.get('groups'):
+                    self.readonly = False
+                    self._authenticated = True
+        except:
```
[the remark this answers is not in the document]
Comment (garberg) [reply nipap/nipap/authlib.py:375]: I must say  agree with Hound here - a bare `except:` is a bit wide. Please add a bit more specific exceptions and more helpful log messages.

File: nipap/nipap/rest.py [hunk 4]
```diff
@@ -45,6 +46,10 @@ def combine_request_args():
 
             if request_nipap_fullname:
                 args['auth'].update({'full_name': request_nipap_fullname})
+        elif request_authorization_header and \
+                request_authorization_header.startswith('Bearer'):
+            authoritative_source = {'auth': {'authoritative_source': 'jwt'}}
```
Comment (garberg) [inline nipap/nipap/rest.py:51]: The authoritative source is used to describe what system is the authoritative source of the data stored in NIPAP (such as a prefix) and doesn't really have that much to do with authentication. That shouldn't need any special handling for JWT auth.

Comment (DanielTheNipapHacker) [reply nipap/nipap/rest.py:51]: Är inte riktigt med på denna. Den behövs för att inte köra ut på rad 148.

**Review by DanielTheNipapHacker — COMMENTED** [review]
Discussed fixes fixed

**Comment by DanielTheNipapHacker** [comment]
Pushed proposed fixes for discussed issues

**Comment by garberg** [comment]
Merged in #1320. Thanks!

**Pull request closed without merging**